### ISSUE TITLE
At line 6925 of file module_ra_rrtm.f90 (unit = 10, file = 'RRTM_DATA') Fortran runtime error: End of file

### ISSUE THREAD

**Issue by wjlyx** [DOI-USGS/COAWST#247]
When I encountered this problem while running the wrf-roms coupling, I tried searching for solutions to similar issues in previous questions. However, it seems that a solution is not readily available. Can you please help me find a way to solve it?
[61676err.txt](https://github.com/DOI-USGS/COAWST/files/15004919/61676err.txt)
[61676out.txt](https://github.com/DOI-USGS/COAWST/files/15004920/61676out.txt)

**Comment by jcwarner-usgs** [comment]
i have seen this error.
https://github.com/DOI-USGS/COAWST/issues/15
when you built the wrf+roms, can you send me that build text.  not the build_coawst.sh, but the text while it builds.
you can do
scrip build.log
./build_coawst.sh -j 5
exit
and then send the build.log here.

**Comment by wjlyx** [comment]
This is my build.log file, and the build did not complete. I didn't notice this before.
[build.txt](https://github.com/DOI-USGS/COAWST/files/15018763/build.txt)

**Comment by wjlyx** [comment]
> i have seen this error. #15 when you built the wrf+roms, can you send me that build text. not the build_coawst.sh, but the text while it builds. you can do scrip build.log ./build_coawst.sh -j 5 exit and then send the build.log here.

wrf+roms
[build.txt](https://github.com/DOI-USGS/COAWST/files/15020282/build.txt)

**Comment by jcwarner-usgs** [comment]
ok. you selected gfortran for WRF
SFC             =       gfortran 
DM_FC           =       mpif90 -f90=$(SFC)
BYTESWAPIO      =       -fconvert=big-endian -frecord-marker=4

and your build for roms is also using mpif90.
So in the ROMS/Compilers/Linux-gfortran.mk
(i think you are using an  older version)
but you should have on line 41
           FFLAGS := -frepack-arrays -fallow-argument-mismatch
and you need to 
also add the line
          FFLAGS += -frecord-marker=4 -fconvert=big-endian
you can add that right below the other one so it could look like:

               FC := gfortran
           FFLAGS := -frepack-arrays -fallow-argument-mismatch
          FFLAGS += -frecord-marker=4 -fconvert=big-endian


and try that.

**Comment by HathewayWill** [comment]
@jcwarner-usgs @wjlyx 

I can confirm that this change works when modifying the COAWST/Compilers/Linux-gfortran.mk file.

Simple way to do it is using a sed command like this.


```
        cd "${COAWST_FOLDER}"/COAWST/Compilers
	sed -i '41a\      FFLAGS += -frecord-marker=4 -fconvert=big-endian' Linux-gfortran.mk
```

**Comment by ImechVN** [comment]
I used Claude Code to try and fix the error, and here are the results.
Root cause found — this is a compile-flag / endianness problem, not a corrupt or missing `RRTM_DATA` file.

I hit this exact error on COAWST v3.8 (WRF + ROMS + SWAN, Sandy test case) and tracked it down. Posting the full analysis since this issue and #104 are both still open.

## TL;DR — the fix

```bash
mpirun -np 3 -x GFORTRAN_CONVERT_UNIT=big_endian ./coawstM Projects/Sandy/coupling_sandy.in
```

The `-x` matters. Without it only the launching rank gets the variable and the other ranks still fail.

## Why checking the file always comes back clean

`RRTM_DATA` is a big-endian Fortran unformatted file. Its first 4 bytes are `00 00 4d 80`:

- read as big-endian → `19840`, the correct length of the first record
- read as little-endian → `2152300544`, far past EOF → `End of file` on the very first `READ`

You can check your own copy:

```bash
od -A d -t x1 -N 4 RRTM_DATA
```

The file is fine. The reader is in the wrong byte order.

## Why WRF's own `-fconvert` flag doesn't save you under COAWST

WRF *is* compiled with `-fconvert=big-endian -frecord-marker=4` (see `WRF/configure.wrf`). But gfortran implements `-fconvert` through a `_gfortran_set_options` call that the compiler emits **only in the main program unit** — not in every object file.

Under COAWST the main program is not WRF's. It is `Master/master.F`, built by the ROMS makefile (`Compilers/Linux-gfortran.mk`), which carries no `-fconvert`.

Quick check on my own build log:

```
grep -c "fconvert=big-endian" build.log   # WRF section: 27
                                          # ROMS section: 0
```

So at runtime the conversion mode is native little-endian, and every big-endian unformatted read WRF performs fails — with a perfectly valid data file. That is why file inspection never leads anywhere.

## Permanent alternative, if you'd rather not rely on an environment variable

Add `-fconvert=big-endian -frecord-marker=4` to the compile line for **`master.f90` only** in `Compilers/Linux-gfortran.mk`. The main program unit is the only place the flag needs to appear.

## Scope warning

`GFORTRAN_CONVERT_UNIT` affects all unformatted I/O in the process. In my configuration that is safe:

- ROMS uses NetCDF throughout (C library, unaffected)
- SWAN uses text I/O
- the only other unformatted I/O I found is `Master/get_numww3_grids.F`, which belongs to WW3 and was disabled in my build

If you have WW3 enabled, scope it per unit instead — e.g. `GFORTRAN_CONVERT_UNIT='big_endian:10'`.

## What does *not* work

Switching to `RRTM_DATA_DBL`. That file is for builds with `-fdefault-real-8`. With `RWORDSIZE = 4` and `PROMOTION` commented out in `configure.wrf`, the single-precision `RRTM_DATA` is the correct file.

## Environment

- Ubuntu 24.04, gfortran 13, OpenMPI
- COAWST v3.8 (`1ccc0b70b061146b4ba631ecd9f6db094c8b65c7`), WRF 4.8.0
- Sandy test case: ROMS 3 nested grids + SWAN 3 grids + WRF 2 domains, `mpirun -np 3`

After this fix the case ran the full 48 simulated hours and finished cleanly with `d01 ... wrf: SUCCESS COMPLETE WRF` and `ROMS: DONE`, producing valid output (sea level −2.45 → 1.16 m, significant wave height 0 → 7.50 m, and non-zero `Uwind` confirming the WRF → ROMS coupling was actually transferring data).

The same fix should apply to #104.

**Comment by jcwarner-usgs** [comment]
thanks for the additional info !
